api/base/pagination.py

```diff
@@ -175,3 +178,136 @@ def get_paginated_response(self, data):
+        self._count = self.object_list['aggs']['total']
+        return self._count
+    count = property(_get_count)
+
+    def page(self, number):
+        number = self.validate_number(number)
+        results = self.object_list['results']
+        items = [
+            self.load_obj(result.get('_id'), result.get('_type'))
+            for result in results
+        ]
+        return self._get_page(items, number, self)
+
+
+class SearchModelPaginator(SearchPaginator):
+
+    def __init__(self, object_list, per_page, model):
+        super(SearchModelPaginator, self).__init__(object_list, per_page)
+        self.model = model
+
+    def page(self, number):
+        number = self.validate_number(number)
+        results = self.object_list['results']
+        items = [
+            self.model.load(result.get('_id'))
+            for result in results
+        ]
+        return self._get_page(items, number, self)
+
+
+class SearchPagination(JSONAPIPagination):
+
+    def __init__(self):
+        super(SearchPagination, self).__init__()
+        self.paginator = None
+
+    def paginate_queryset(self, queryset, request, view=None):
+        page_size = self.get_page_size(request)
+        if not page_size:
+            return None
+
+        self.paginator = SearchPaginator(queryset, page_size)
+        model = getattr(request.parser_context['view'], 'model_class', None)
+        if model:
+            self.paginator = SearchModelPaginator(queryset, page_size, model)
+
+        page_number = request.query_params.get(self.page_query_param, 1)
+        if page_number in self.last_page_strings:
+            page_number = self.paginator.num_pages
+
+        try:
+            self.page = self.paginator.page(page_number)
+        except InvalidPage as exc:
+            msg = self.invalid_page_message.format(
+                page_number=page_number, message=six.text_type(exc)
+            )
+            raise NotFound(msg)
+
+        if self.paginator.num_pages > 1 and self.template is not None:
+            # The browsable API should display pagination controls.
+            self.display_page_controls = True
+
+        self.request = request
+        return list(self.page)
+
+    def get_search_field_url(self, query, field):
+        view_name = 'search:search-{}'.format(field)
+        return absolute_reverse(
+            view_name,
+            query_kwargs={
+                'q': query
+            }
+        )
+
+    def get_search_field_total(self, field):
+        return self.paginator.object_list['counts'].get(field, 0)
+
+    def get_response_dict(self, data, url):
+        if isinstance(self.paginator, SearchModelPaginator):
+            return super(SearchPagination, self).get_response_dict(data, url)
+        else:
+            query = self.request.query_params.get('q', '*')
+            return OrderedDict([
+                ('data', data),
+                ('search_fields', OrderedDict([
+                    ('files', OrderedDict([
```

Comment:
I forgot to warn you this morning, but the links here are going to need to be the same general structure as our relationship links. We're doing the pagination links wrong, but we can't change them until we have API Versioning in. But it'll probably look something like:

```
links:
    first
    last
    prev
    next
    meta:
        total
        per_page
    users:
        related:
            href
            meta:
                total
    projects:
        related:
            href
            meta:
                total
    components:
        related:
            href
            meta:
                total
```